### ISSUE TITLE
Pagination in 'cms:folders' and 'cms:no_results'

### ISSUE THREAD

**Issue by trendoman** [CouchCMS/CouchCMS#140]
The folders being paginated with 'cms:folders' continue to show up after the last page.
```html
<cms:folders masterpage='index.php'
    root='world'
    paginate='1'
    hierarchical='1'
    limit='1'
    >
    <cms:dump />
</cms:folders>
```
The code above would continue to show folders, even when the URL is http://example.com/index.php?pg=999

#### I suggest to follow the suit of 'cms:pages' tag &mdash;

The pages being paginated with 'cms:pages' never show up after the last page.
```html
<cms:pages masterpage='index.php' 
    limit='1' 
    paginate='1'>
    <cms:dump />
</cms:pages>
```
The code above would **expectedly** cease to show pages when the URL is http://example.com/index.php?pg=999

Also, both tags are thought to support 'cms:no_results', however it is currently only working with 'cms:pages'.

**Comment by kksidd** [comment]
Regarding the first issue you reported where records 'continue to show up after the last page' -
you'll notice that Couch actually figures out that the requested page number is higher than the last available but then goes on to set it to the later -
`if( $pgn_pno>$total_pages && $total_pages>0 ) $pgn_pno=$total_pages;`

As to why it does so, you'll have to keep in mind that this paginated listing of 'folders' and 'nested_pages' like normal 'pages' was primarily introduced for handling them in the admin-panel.

Allow me to explain -
These two hierarchical entities, unlike normal 'pages', can possibly have multiple ancestors.
So, suppose on the last page of a paginated set, we have only a single folder/nested_page to show but that has, say, 7 ancestors; the admin-panel actually first displays the ancestors and then the solitary folder (i.e. shows 8 entities instead of 1).

Now suppose the user deletes any of the 8 folders shown on the last page. When the operation completes, the paginated set gets reduced by one and so the current page becomes higher than the last available.

If the display logic of normal 'pages' (and which you are suggesting) is used, the user will be greeted by a blank page and it would appear to him that all the 8 folders that were previously visible are gone!
To avoid this scenario, Couch silently adjusts the page to show to be the last available and things look normal.

In short, it was a concious decision and, while things could have been handled differently, that is how they stand currently and changing anything would break backward-compatibility.
The behaviour you desire, as you have shown, is fairly easy to simulate but if you must have it in code you may introduce a new param with <cms:folders> and <cms:nested_pages> to override the default behavior.
Please get a PR and I'll merge it.

Now for the second issue (<cms:no_results> not working) -
I tested and, indeed, it is an issue while using 'root' and 'childof' params.
Have commited a fix - https://github.com/CouchCMS/CouchCMS/commit/560d9e99182e0e3af899eca3040d4f698e590ec7

Please test and let me know it it works.

Thanks.

**Comment by trendoman** [comment]
Doesn't work for me, unfortunately. Tag 'no_results' doesn't kick in.

**Comment by trendoman** [comment]
> Doesn't work for me, unfortunately. Tag 'no_results' doesn't kick in.

Ok. I had folders, my bad. However &mdash;

If someone creates a pagination for blog categories - pagination becomes a carousel, **stuck on the last item** _by default_! 

---

It's really hard to question the authorities, but no matter how lengthy is the explanation and references to 'old times' - pagination **is** broken in folders. Must be fixed without questions. Otherwise tags simply do not work universally. I don't buy it. We can have a vote in forum.

**Comment by trendoman** [comment]
> As to why it does so, you'll have to keep in mind that this paginated listing of 'folders' and 'nested_pages' like normal 'pages' was primarily introduced for handling them in the admin-panel.
> 

Does no one list folders aka blog categories or product categories **on the frontend** paginated? The problem here is that pagination is enabled alongside **hierarchical** parameter. 

Kindly allow paginate folders on the frontend just like _expected_ and it will be fine. Presumably -- 
```
<cms:folders masterpage='index.php'
    root='world'
    paginate='1'
    limit='1'
    >
</cms:folders>
```
And allow :no_results to actually kick in, when there are no more items to enumerate.

Outrageous, if you ask me.
